rust/lance/src/index/vector/ivf.rs

```diff
@@ -2546,6 +2515,116 @@ mod tests {
         }
     }
 
+    #[tokio::test]
+    async fn test_create_ivf_pq_cosine() {
+        run_ivf_pq_cosine_test(false).await;
+    }
+
+    #[tokio::test]
+    async fn test_create_ivf_pq_v3_cosine() {
+        run_ivf_pq_cosine_test(true).await;
+    }
+
+    #[tokio::test]
+    async fn test_build_ivf_model_l2() {
+        let test_dir = tempdir().unwrap();
+        let test_uri = test_dir.path().to_str().unwrap();
+
+        let (dataset, _) = generate_test_dataset(test_uri, 1000.0..1100.0).await;
+
+        let ivf_params = IvfBuildParams::new(2);
+        let ivf_model = build_ivf_model(&dataset, "vector", DIM, MetricType::L2, &ivf_params)
+            .await
+            .unwrap();
+        assert_eq!(2, ivf_model.centroids.as_ref().unwrap().len());
+        assert_eq!(32, ivf_model.centroids.as_ref().unwrap().value_length());
+        assert_eq!(2, ivf_model.num_partitions());
+
+        // All centroids values should be in the range [1000, 1100]
+        ivf_model
+            .centroids
+            .unwrap()
+            .values()
+            .as_primitive::<Float32Type>()
+            .values()
+            .iter()
+            .for_each(|v| {
+                assert!((1000.0..1100.0).contains(v));
+            });
+    }
+
+    #[tokio::test]
+    async fn test_build_ivf_model_cosine() {
+        let test_dir = tempdir().unwrap();
+        let test_uri = test_dir.path().to_str().unwrap();
+
+        let (dataset, _) = generate_test_dataset(test_uri, 1000.0..1100.0).await;
+
+        let ivf_params = IvfBuildParams::new(2);
+        let ivf_model = build_ivf_model(&dataset, "vector", DIM, MetricType::Cosine, &ivf_params)
+            .await
+            .unwrap();
+        assert_eq!(2, ivf_model.centroids.as_ref().unwrap().len());
+        assert_eq!(32, ivf_model.centroids.as_ref().unwrap().value_length());
+        assert_eq!(2, ivf_model.num_partitions());
+
+        // All centroids values should be in the range [1000, 1100]
+        ivf_model
+            .centroids
+            .unwrap()
+            .values()
+            .as_primitive::<Float32Type>()
+            .values()
+            .iter()
+            .for_each(|v| {
+                assert!(
+                    (-1.0..1.0).contains(v),
+                    "Expect cosine value in range [-1.0, 1.0], got: {}",
+                    v
+                );
+            });
+    }
+
+    #[tokio::test]
+    async fn test_create_ivf_pq_dot() {
+        run_ivf_pq_dot_test(false).await;
+    }
+
+    #[tokio::test]
+    async fn test_create_ivf_pq_v3_dot() {
+        run_ivf_pq_dot_test(true).await;
+    }
```

Comment:
Instead of making separate test functions, could we parametrize the test functions with rstest?